lisp/ess-utils.el

```diff
@@ -92,12 +92,61 @@ Used in noweb modes.")
 
 ;;*;; elisp tools
 
+(defun ess-next-code-line (&optional arg skip-to-eob)
+  "Move ARG lines of code forward (backward if ARG is negative).
+Skips past all empty and comment lines. Default for ARG is 1.
+Don't skip the last empty and comment lines in the buffer unless
+SKIP-TO-EOB is non-nil. On success, return 0. Otherwise, go as
+far as possible and return -1."
+  (interactive "p")
+  (or arg (setq arg 1))
+  (beginning-of-line)
+  (let ((pos (point))
+        (n 0)
+        (inc (if (> arg 0) 1 -1)))
+    (while (and (/= arg 0) (= n 0))
+      (setq n (forward-line inc)); n=0 is success
+      (if (not (fboundp 'comment-beginning))
+          (while (and (= n 0)
+                      (looking-at "\\s-*\\($\\|\\s<\\)"))
+            (setq n (forward-line inc)))
+        (comment-beginning)
+        (beginning-of-line)
+        (forward-comment (* inc (buffer-size))) ;; as suggested in info file
+        )
+      (if (or skip-to-eob
+              (not (looking-at ess-no-skip-regexp))) ;; don't go to eob or whatever
+          (setq arg (- arg inc))
+        (goto-char pos)
+        (setq arg 0)
+        (forward-line 1));; stop at next empty line
+      (setq pos (point)))
+    (goto-char pos)
+    n))
+
 (defun ess-goto-line (line)
   (save-restriction
     (widen)
     (goto-char (point-min))
     (forward-line (1- line))))
 
+(defun ess-skip-thing (thing)
+  "Leave point at the end of THING.
+THING can be 'function, 'paragraph, or 'line."
+  (cond
+   ((eql thing 'line) (goto-char (line-end-position)))
+   ((eql thing 'paragraph) (forward-paragraph))
+   ((eql thing 'function) (goto-char (cadr (ess-end-of-function nil 'no-error))))))
+
+(defun ess-step-line (&optional skip)
```

Comment:
I like the name since it's called from all the `eval-and-step` functions. The naming is consistent with those.

Reply:
> That's not a good reason

Well I think it is.

> This function does much more than those -step functions require

It skips something then steps to the next line. I suggested `ess-step-and-skip` or something like that but you didn't like that.

What would you like to name it?

Reply:
>  I suggested ess-step-and-skip

It's the other way around `ess-skip-and-step`. You see, even you are confused. Those skip and step are close relatives and we don't want people to deal with the marginal difference in semantics, but I am repeating myself. 

> What would you like to name it?

I would like to merge it into `ess-skip-thing` as proposed earlier.  It would make the code using the function more readable. 

Proliferation of new functions with marginal value is something that we want to avoid in the future. The mess that we are in is mostly because different people (myself including) were adding over the years whatever they thought is useful without consideration for the big picture and overall design.

Reply:
I'm not sure how that's any different than the current situation, but sure. How do you want to merge it in? Like this?

```diff
diff --git a/lisp/ess-utils.el b/lisp/ess-utils.el
index 2f480ae5..e0e27c50 100644
--- a/lisp/ess-utils.el
+++ b/lisp/ess-utils.el
@@ -132,18 +132,12 @@ far as possible and return -1."
     (forward-line (1- line))))
 
 (defun ess-skip-thing (thing)
-  "Leave point at the end of THING.
+  "Leave point at the end of THING then go to the \"next\" line.
 THING can be 'function, 'paragraph, or 'line."
   (cond
    ((eql thing 'line) (goto-char (line-end-position)))
    ((eql thing 'paragraph) (forward-paragraph))
-   ((eql thing 'function) (goto-char (cadr (ess-end-of-function nil 'no-error))))))
-
-(defun ess-step-line (&optional skip)
-  "Step forward to the \"next\" line.
-Depending on `ess-eval-empty', step one line or to the next code
-line. When given, SKIP gets passed to `ess-skip-thing'."
-  (when skip (ess-skip-thing skip))
+   ((eql thing 'function) (goto-char (cadr (ess-end-of-function nil 'no-error)))))
   (if (or ess-eval-empty
           (and (fboundp 'ess-roxy-entry-p)
                (ess-roxy-entry-p)))
```